### PR TITLE
[ClavaLaraApi] Adds the function outliner to Clava's mainline

### PR DESCRIPTION
**Pull request by tiagolascasas**
Adds a function outliner (i.e., it extracts a code region into its own function, and replaces the region by a call to that function) to Clava. All public methods are documented.

Usage example:

```
laraImport("clava.code.Outliner");

const outliner = new Outliner();
outliner.outline(begin, end);
```

where `begin` and `end` are the statements that start and end the region to be outlined. They must be within the same scope (that and further restrictions are verified by the outliner itself, and can also be checked manually by using the `checkOutline()` method).

Test with the following code:
[ClavaFunctionOutliner.zip](https://github.com/specs-feup/clava/files/11503597/ClavaFunctionOutliner.zip)

### CODE REVIEW

File: ClavaLaraApi/src-lara-clava/clava/clava/code/Outliner.js [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+        }
+        // no need to handle global vars - they are not parameters
+
+        const args = [];
+        for (const param of fun.params) {
+            for (const decl of decls) {
+                if (decl.name === param.name) {
+                    const ref = ClavaJoinPoints.varRef(decl);
+
+                    if (param.type.joinPointType === "pointerType" && ref.type.joinPointType === "builtinType") {
```
[the remark this answers is not in the document]
Comment (tiagolascasas) [reply ClavaLaraApi/src-lara-clava/clava/clava/code/Outliner.js:259]: Done

File: ClavaLaraApi/src-lara-clava/clava/clava/code/Outliner.js [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+    }
+
+    #createParams(varrefs) {
+        const params = [];
+
+        for (const ref of varrefs) {
+            const name = ref.name;
+            const varType = ref.type;
+
+            if (["arrayType", "adjustedType", "pointerType"].includes(varType.joinPointType)) {
```
[the remark this answers is not in the document]
Comment (tiagolascasas) [reply ClavaLaraApi/src-lara-clava/clava/clava/code/Outliner.js:335]: Done

File: ClavaLaraApi/src-lara-clava/clava/clava/code/Outliner.js [hunk 3]
```diff
@@ -0,0 +1,464 @@
+"use strict";
+
+laraImport("weaver.Query");
+laraImport("clava.ClavaJoinPoints");
+laraImport("clava.ClavaType");
+
+class Outliner {
+    static #GLOBAL_OUTLINE_FUN_ID = 1;
```
[the remark this answers is not in the document]
Comment (tiagolascasas) [reply ClavaLaraApi/src-lara-clava/clava/clava/code/Outliner.js:8]: Done

File: ClavaLaraApi/src-lara-clava/clava/clava/code/Outliner.js [hunk 4]
```diff
@@ -154,7 +153,7 @@ class Outliner {
 
         // actions before the function call
         const type = returnStmts[0].children[0].type;
-        const id = Outliner.#getUniqueVarID();
+        const id = IdGenerator.next();
```
Comment (joaobispo) [inline ClavaLaraApi/src-lara-clava/clava/clava/code/Outliner.js:166]: Just a detail, if you want to customize the ID, and have a separate numbering for functions and variables, you can pass a string as argument, `IdGenerator.next("var_")` will return `var_0`, `var_1`, etc

Comment (tiagolascasas) [reply ClavaLaraApi/src-lara-clava/clava/clava/code/Outliner.js]: That's nice, I've updated accordingly.

**Review by joaobispo — APPROVED** [review]

**Pull request merged**